lsコマンド1プログラムを作成

sochi419 opened the pull request.

sochi419 committed:
Add ls

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,111 @@
+
+class Ls
+  # 要素数が4以下の場合、formula1を適用する。
+  def way1
```
`way1`, `way2`, `way3`.. みたいなメソッド名は、何をするメソッドなのかが全く分からないのであまり良くないですね。理想はメソッド名だけ読めば何をするのか理解できることです。
とりあえず、全てのメソッドに何かしら意味のある名前をつけてみてください。

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,111 @@
+end
+
+ls = Ls.new
+arr = Dir.glob('*')
```
`Ls`クラスの中でも`Dir.glob('*')`しているのに、ここでも実行していますね。無駄な処理になってしまっているのでやめましょう。
個人的には、今`Ls`クラスの外に書いてあるような処理も`Ls`クラスの責務として持たせた方がいいと思います。そうすれば`Dir.glob('*')`するのもLsクラスの中で一度だけで済みます。

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,111 @@
+ls = Ls.new
+arr = Dir.glob('*')
+# 列数nとする。
+n = 3
```
プログラムを通して不変な値ですし、これはいわゆる設定値のようなものなので、定数にしてしまうと良いと思います。そうすることで、

> 「3列から100列に変えてください」とあとから言われても必要最小限の変更で対応できるようなロジックにしておくこと

が実現しやすくなります。（ここを変えればいいんだな、と分かりやすくなる）

summertree128 replied:
あと、`n`ではなくて意味のある定数名にしてください。

summertree128 replied:
ちなみにですが、定数はプログラムの一番上あたりに置くのが慣習なので、そのようにしてください〜

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,111 @@
+class Ls
+  # 要素数が4以下の場合、formula1を適用する。
+  def way1
+    arr = Dir.glob('*')
```
`arr`というのは型の情報であって、中に何が入っているかの情報がありませんね。
中に何が入っているのかがわかる変数名にしましょう。また、変数名に型の情報を持たせるのはあまり良くないのでなるべく避けましょう。

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,111 @@
+  def way1
+    arr = Dir.glob('*')
+    # 要素数m 列数nとする。
+    m = arr.size
```
コメントを書くのではなく、そのコメントのような処理をしていることが一目で分かるような変数名・メソッド名にしましょう。

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,111 @@
+    m = arr.size
+    # 出力位置を統一するため、変数max, empty1を定義
+    max = arr.max_by(&:length).length
+    empty1 = ' ' * (max - arr[0].length)
```
基本的に`hoge1` `hoge2`みたいな変数名が適切であることはほとんどないので、数値で連番をつけるのはやめましょう〜

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,111 @@
+  print ls.way2.to_s
+else
+  print ls.way3.to_s
+end
```
ファイルの数で実行するメソッドをまるっと分岐するのではなく、

- なるべく分岐せずに全てパターンに対応できるようなロジックにする
- やむなく分岐が必要な際は、分岐する部分を最小限にする（最初から分岐するのではなく、必要なところで分岐する）

ようにしてみてください。おそらく、このような設計だと後から追加の要件が来たときにとてつもなく辛いことになります。（具体的にはls 2以降がえらいことになります）

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,111 @@
+      print "\n"
+      print (arr[1]).to_s
+      print "\n"
+    end
```
やろうとしていることは良いと思いますが、要素数で分岐するのではなく、例えば良い感じの配列を作って、要素数が何であってもうまく出力できるような処理にしてみて下さい。（つまり`case~when`をなくす）
それを突き詰めていくと、`way1`, `way2`, `way3`のようなメソッドレベルでの分岐も不要になるはずです。

summertree128 reviewed (commented):
レビューしました！

動作的には良い感じですし、やろうとしていることも理解できます。
ただ、こういう形でメソッドを分割したり条件分岐していると、柔軟性のない設計になってしまいます。仕様変更や機能追加がかなり難しくなってしまうので、もう少し分岐やコードの重複が少ないような形にリファクタリングしていきたいです。

というわけで、たくさんコメントしたのでご確認をお願いしますー。

sochi419 committed:
second commit

sochi419 commented:
指摘点を踏まえて、提出したコードを変更しました。
↓主な変更点↓
① 変数名をなるべくわかりやすい名前にする。
② なるべく分岐せずに全てパターンに対応できるようなロジックにする。

以前のコードの大部分を変更したので、確認が大変だとは思いますが、
よろしくおねがいします🙏

sochi419 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,32 @@
+      frames = @elements.each_slice(@elements.size / @column).to_a
+    else
+      frames = @elements.each_slice(@elements.size / @column + 1).to_a
+      frames.last << nil until frames.first.size == frames.last.size
```
後述する[transposeメソッド](https://docs.ruby-lang.org/ja/latest/method/Array/i/transpose.html)を使うために、
要素数が足りない配列にnilを代入しています。

summertree128 replied:
説明ありがとうございます。「読み手にとって意図が分かりにくそうな部分にコメントをつける」というのは良いプラクティスだと思います。
せっかくなのでコード内コメントとして記述しておいてもらうとより良いかなと思います！

sochi419 replied:
コメントアウトとして、コード内に記載しました。

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+class Ls
+  def instance
```
「インスタンスを生成する時に実行する処理」というような意味合いで `instance` というメソッド名なのでしょうか？
もしそうであれば、それはコンストラクタが担うべき処理なので、コンストラクタ内に処理を書いて下さい。

参考：https://docs.ruby-lang.org/ja/latest/method/Object/i/initialize.html

sochi419 replied:
initializeメソッドに名前を変更しました。

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,32 @@
+
+class Ls
+  def instance
+    @elements = Dir.glob('*')
```
`@elements`という名前はちょっと抽象的すぎて「この変数の中に何が入っているのか」がわかりづらいので、もう少し具体的で変数の中身が推測できるような名前を考えてみてもらいたいです。

sochi419 replied:
"ディレクトリの要素群"という意味で、"directory_elements"と命名しました。

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,32 @@
+class Ls
+  def instance
+    @elements = Dir.glob('*')
+    @column = 3
```
`@column`は基本的にこのプログラム中で不変な値なので、定数にしたほうがいいかなと思いました。

sochi419 replied:
定数としてCOLUMNと命名し、クラスの外に配置しました。

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,32 @@
+    @column = 3
+  end
+
+  def make_arr
```
`make_arr`だと何をするメソッドなのか少しわかりづらいですね。もう少し具体的な名前にしてもらいたいです。
シンプルに考えるならば、`@frames`を作るためのメソッドなので「フレームを作る」を英語にした感じの名前でしょうか。
ただ、`@frames`もちょっと意味が分かりづらいので、もう少し本質的な意味を表している名前にしてもらいたいです。フレームってなんだっけ…という感じなので、「この変数には何が入っているのか」を誰かに説明するとしたらなんと説明するかを考えてみてください。それを英語にしたものが変数名になるといいのかなと思います。

sochi419 replied:
- make_arrメソッドを消して、initialized メソッドの中に、
make_arrメソッドの内容を含みました。
- "区分けされた要素群"という意味で、"divided_elements"と命名しました。

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,32 @@
+  end
+
+  def output
+    make_arr
```
`make_arr`も初期化処理っぽい感じがするので、コンストラクタの中で実行しちゃってもいいかもしれませんね。

sochi419 replied:
make_arrメソッドを消して、initialized メソッドの中に、
make_arrメソッドの内容を含みました。

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,32 @@
+
+  def output
+    make_arr
+    reverse = @frames.transpose
```
`reverse`だと中に何が入っているのか名前から全く推測できないので、もう少し意味のある名前にしてもらいたいです。
あまりいいのが思いつかない場合は`transposed_frames`とかでもいいと思います。（ただし`@frames`の方が変わったらこちらも併せて変えてください）

sochi419 replied:
"整列された要素群"という意味で、"sorted_elements"と命名しました。

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,32 @@
+    reverse = @frames.transpose
+    reverse.each do |x|
+      x.delete(nil)
+      x.map { |n| n << (' ' * (@elements.max_by(&:length).length - n.length)).to_s }
```
`@elements.max_by(&:length).length`が相当な回数実行されていて、少し無駄な処理になってしまっていますね。
実際には、この値はループの中で変化したりしないので、ループの前に一度だけ実行すれば十分なはずです。

sochi419 replied:
`@elements.max_by(&:length).length`
をループ前に処理するコードに書き換えました。

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,32 @@
+    reverse.each do |x|
+      x.delete(nil)
+      x.map { |n| n << (' ' * (@elements.max_by(&:length).length - n.length)).to_s }
+      print "#{x.join(' ')}\n"
```
このままでも特に問題があるわけではないんですが、常に改行するのであれば`puts`を使った方が明快でいいかなと思いました。わざわざ改行コードを記述する必要もなくなりますので。

sochi419 replied:
putsに変更しました。

summertree128 reviewed (commented):
相当変わりましたね！！全体的な設計としては前回のコードとは比べ物にならないくらいよくなったと思います。記述量も劇的に減っていて素晴らしいと思います。

変数やメソッドの命名あたりを中心にコメントしていますのでご確認をお願いしますー。

sochi419 committed:
third commit

sochi419 commented:
指摘点を修正しました。
確認よろしくおねがいします🙏

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -1,32 +1,31 @@
-    if (@elements.size % @column).zero?
-      frames = @elements.each_slice(@elements.size / @column).to_a
+  def initialize
+    @directory_elements = Dir.glob('*')
```
変数名としてはかなりベターになりましたね。ただ、じつは`elements`という言葉にはほとんど情報量がないような気がしています。（配列である以上、"要素"が複数入っていることは自明なので）
もうちょっと具体的にいうと、配列の１要素が何を表しているかというのを考えた時、今回で言えば`directory`ですよね。であれば、それを複数形にした`directories`というのが変数名として必要十分かなと思います。
他の配列も同じように変数名を考え直してみてください。

sochi419 replied:
なるほど、要素をそのまま英語にして、elementsと命名していました...
確かにdirectoriesの方が、伝わりやすいですね !

変数名を、elements→directoriesに変更しました。

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -1,32 +1,31 @@
+      end
     end
-    @frames = frames
+    @divided_elements = divided_elements
```
すみません、ちょっとコンストラクタが大きくなり過ぎてしまいましたね。
基本的にコンストラクタはもう少しシンプルであって欲しい気がしますね。せめてこの辺の処理をメソッドとして切り出して、こんな感じになるともう少し見やすいと思います。

```ruby
# 注：divide_elements という名前のメソッドを切る想定です
@divided_elements = divide_elements
```

また、上にもコメントしたように`elements`というのはあまり情報量がないので、もう少し具体的な情報がわかるような名前にしてもらいたいです。

sochi419 replied:
initializeメソッドをシンプルにするために、divided_directories メソッドを作成し、
initializeメソッド内部にあった処理の一部を移動しました。

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -1,32 +1,31 @@
   end
 end
 
+COLUMN = 3
```
定数は習慣的にプログラムの初めの方に置くことが多いので、クラスの宣言のすぐ下(4行目あたり)に書いておくといいと思います。

sochi419 replied:
クラス宣言のすぐ下で、定数を定義しました。

summertree128 reviewed (commented):
いい感じですね！
あと少しだけ気になる点があるのでコメントしています。ご確認お願いしますー

summertree128 commented:
あと、コミットメッセージにはもう少し具体的に「どういう変更をしたか」を書いてもらいたいです！
参考：https://docs.komagata.org/5744

sochi419 committed:
FIX improvements

Fixed the following points
1. Change variable name
2. Method simplification
3. Change the position of the constant definition

sochi419 commented:
指摘点の修正をしました!
また、コミットメッセージを具体的にしました。

確認よろしくお願いします🙏

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,35 @@
+    @directories = Dir.glob('*')
+  end
+
+  def divided_directories
```
メソッド名は原則的に動詞で始めましょう。

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,35 @@
+    @directories = Dir.glob('*')
+  end
+
+  def divided_directories
```
メソッド名は原則的に動詞で始めましょう。

sochi419 replied:
メソッド名を変更しました。
- divided_directories → divide_directories
- output → sort_directories

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,35 @@
+        divided_directories.last << nil
+      end
+    end
+    @divided_directories = divided_directories
```
このインスタンス変数とローカル変数が両方ある意味があまりないような気がするので、どちらかだけにした方がいいと思います。具体的には

- 11行目や13行目の時点で最初からインスタンス変数を更新する
- このメソッドはインスタンス変数を使わず、 `divided_directories`を返すようにする

のどちらかがいいかなと思います。個人的には、これは`@directories`があればそこから導出できる値なのでインスタンス変数として保持する必要性は薄いかなと思いますので後者の方が好みではありますが、どちらでも大丈夫です。

sochi419 replied:
> このメソッドはインスタンス変数を使わず、 divided_directoriesを返すようにする

返り値をdivide_directoriesになるよう修正しました。

summertree128 commented on 05.ls/ls_sub.rb:
```diff
@@ -0,0 +1,35 @@
+        divided_directories.last << nil
+      end
+    end
+    @divided_directories = divided_directories
```
このインスタンス変数とローカル変数が両方ある意味があまりないような気がするので、どちらかだけにした方がいいと思います。具体的には
- 11行目や13行目の時点で最初からインスタンス変数を更新する
- このメソッドはインスタンス変数を使わず、 `divided_directories`を返すようにする
のどちらかがいいかなと思います。個人的には`@directories`があればそこから導出できる値なのでインスタンス変数として保持する必要性は薄いかなと思いますので後者の方が好みではありますが、どちらでも大丈夫です。

sochi419 committed:
Fixed the following points
1. Change method name
2. Fix Method "divide_directories"

sochi419 commented:
修正しました！

確認よろしくお願いします🙏

summertree128 reviewed (approved):
レビューOKです！
